fix: pubsub blocking pull integration test (#11025)

alevenberg opened the pull request:
Fixes issue #11025
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11983)
<!-- Reviewable:end -->

coryan commented on google/cloud/pubsub/internal/subscriber_connection_impl.cc:
```diff
@@ -55,16 +55,22 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
-      request, __func__);
-  if (!response) return std::move(response).status();
-  if (response->received_messages_size() != 1) {
+  auto messages_received = 0;
+  StatusOr<google::pubsub::v1::PullResponse> response;
+  while (messages_received == 0) {
+    response = google::cloud::internal::RetryLoop(
+        current.get<pubsub::RetryPolicyOption>()->clone(),
+        current.get<pubsub::BackoffPolicyOption>()->clone(),
+        google::cloud::Idempotency::kIdempotent,
+        [stub = stub_](auto& context, auto const& request) {
+          return stub->Pull(context, request);
+        },
+        request, __func__);
+
+    if (!response) return std::move(response).status();
+    messages_received = response->received_messages_size();
+  }
```
This implies we have a retry loop inside a retry loop.  That seems odd.  If I (or an application developer writes):

```cc
  auto m = client.Pull(..., Options{}.set<pubsub::RetryPolicyOption>(LimitedTimeRetryPolicy(60s)));
```

I expect the client library to block for at most 60s, not for "60s unless no message arrive in which case block for 60s more"?

Maybe something like this instead?

```suggestion
    auto response = google::cloud::internal::RetryLoop(
        current.get<pubsub::RetryPolicyOption>()->clone(),
        current.get<pubsub::BackoffPolicyOption>()->clone(),
        google::cloud::Idempotency::kIdempotent,
        [stub = stub_](auto& context, auto const& request) {
          auto r = stub->Pull(context, request);
          if (!r || !r->received_messages().empty()) return r;
          // No messages received, return a retryable error to
          // try again. This respects the retry policy.
          return google::cloud::internal::AbortedError("no messages received", GCP_ERROR_INFO());
        },
        request, __func__);
```

coryan replied:
I am not sure `AbortedError` is the right way to represent this condition.  We may need a different error code, or a completely different retry loop.  But the point about the double loop stands I think.

alevenberg replied:
Added a custom retry loop.

coryan commented on google/cloud/pubsub/internal/subscriber_connection_impl_test.cc:
```diff
@@ -409,6 +409,52 @@ TEST(SubscriberConnectionTest, Pull) {
   t.join();
 }
 
+TEST(SubscriberConnectionTest, PullReturnsNoMessage) {
```
Consider a test that:

1) Sets the retry policy to tolerate at most N failures.
2) Returns empty messages N times.
3) The response should be an error of some kind.

alevenberg replied:
Added test

alevenberg commented:
Accidentally reset the initial commit (my bad)

Added a new retry loop and test that you suggested. Also, fixed the off by 1 error in the limited count retry policy.

Tried to run on the CI locally, but my resources aren't set up right. Looking into now.

alevenberg commented:
I can't tell if it's an off by 1 error or not. A lot of tests break when I changed the IsExhausted() condition on the limited error count retry policy.  

It looks like there are some other places where we set a error count, but then make one more RPC (like in DefaultRowReaderTest, RetryPolicyExhausted [here](https://github.com/alevenberg/google-cloud-cpp/blob/284eadb43d46927fe81655c986467bc56262b7cb/google/cloud/bigtable/internal/default_row_reader_test.cc#L243))

devbww commented on google/cloud/pubsub/internal/subscriber_connection_impl.cc:
```diff
@@ -53,31 +55,54 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
+  if (!retry_policy->IsExhausted()) {
+    return google::cloud::internal::RetryLoopError("Permanent error in",
+                                                   __func__, last_status);
   }
-  auto received_message =
-      std::move(response->mutable_received_messages()->at(0));
-  auto impl = std::make_unique<pubsub_internal::DefaultPullAckHandler>(
-      background_->cq(), stub_, current, std::move(subscription),
-      std::move(*received_message.mutable_ack_id()),
-      received_message.delivery_attempt());
-  auto message = pubsub_internal::FromProto(
-      std::move(*received_message.mutable_message()));
-  return pubsub::PullResponse{pubsub::PullAckHandler(std::move(impl)),
-                              std::move(message)};
+  return google::cloud::internal::RetryLoopError("Retry policy exhausted in",
+                                                 __func__, last_status);
```
It's a little hard to tell in the diff, but this looks strange on the face of things.
```
  if (!last_status.ok()) return A;
  if (last_status.ok()) return B;
  if (!retry_policy->IsExhausted()) return C;
  return D;
```
Isn't everything after the second statement unreachable?

alevenberg replied:
That's correct. Removed the first if.

Now, it does the following 1) if the last status was ok, it means the rpc returned 0 messages and will give the unavailable error, 2) if the policy is not exhausted, it was a permanent error, 3) return the policy was exhausted.

I tried removing the braces {}, but the formatting is still a little weird. Any suggestions?

devbww replied:
We generally don't remove braces when the `if` covers multiple lines, so I'd only suggest restoring them.

And one more optional nit ... I'd suggest inverting the final `IsExhausted()` check so that your flow is _"2) if the policy is exhausted, return the policy is exhausted, 3) return permanent error"_, just so the exhausted conditional is more directly related to the exhausted error (and eliminating negations is usually a good thing).

alevenberg replied:
Re added the brackets and agreed, swapped the order of the ifs/return

coryan commented on google/cloud/pubsub/internal/subscriber_connection_impl.cc:
```diff
@@ -53,31 +55,51 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
 
   google::pubsub::v1::PullRequest request;
   request.set_subscription(subscription.FullName());
+  // Pubsub returns at most 1 message.
```
nit: maybe "Ask Pub/Sub to return at most 1 message".

coryan commented on google/cloud/pubsub/internal/subscriber_connection_impl.cc:
```diff
@@ -53,31 +55,51 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
+  auto retry_policy = current.get<pubsub::RetryPolicyOption>()->clone();
+  auto backoff_policy = current.get<pubsub::BackoffPolicyOption>()->clone();
+  google::cloud::Status last_status;
+  google::cloud::StatusOr<google::pubsub::v1::PullResponse> response;
```
`response` is only used inside the loop, so maybe move it there and you can use `auto response = ...`.

coryan commented on google/cloud/pubsub/internal/subscriber_connection_impl.cc:
```diff
@@ -53,31 +55,51 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
+    last_status = std::move(response).status();
+    if (!retry_policy->OnFailure(last_status)) {
+      break;
+    };
```
nit: extra semi-colon

coryan reviewed (approved):
I think this is almost ready.  Just a few nits.
